core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java

```diff
@@ -6349,6 +6349,47 @@ private MyProjectRule(Class<? extends Project> clazz,
     String planAfter = NL + RelOptUtil.toString(relAfter);
     getDiffRepos().assertEquals("planAfter", "${planAfter}", planAfter);
   }
+
+  @Test public void testDecorrelateUnion0() {
+    final String sql = ""
+        + "SELECT deptno FROM dept where exists\n"
+        + "(SELECT 1 FROM emp where sal < 100 and emp.deptno=dept.deptno\n"
+        + "UNION ALL\n"
+        + "SELECT 1 FROM emp where sal > 200 and emp.deptno=dept.deptno)";
+    sql(sql)
+        .withLateDecorrelation(true)
+        .withTrim(true)
+        .with(HepProgram.builder().build())
+        .check();
+  }
+
+  @Test public void testDecorrelateUnion1() {
+    // Decorrelation would fail because the input frames of Union
+    // cannot match by row schema.
+    final String sql = "SELECT * FROM emp A where (deptno, sal) in \n"
+        + "(SELECT deptno, sal FROM emp\n"
+        + "UNION ALL\n"
+        + "SELECT deptno, sal FROM emp B where A.mgr = B.empno)";
```

Comment:
Isn't this a valid case to decorrelate ?

Reply:
Thanks for comment @danny0405 
Yes, I believe so, it's a valid case.
But based on current implementation/mechanism of RelDecorrelator, oldToNewOutputs and corDefOutputs should be derived and transferred from bottom to up properly.
In such case, we cannot derive the `oldToNewOutputs` for Frame of Union and this PR doesn't target to resolve such issue.

Reply:
Thanks for the explain, this PR has too much limitation if we do not solve such pattern, i kind of didn't like the idea that because of the current implementation, we can not make the PR complete, if that is the case, then we should fire another issue to refactor the Union frame to support such decorrelation and resolve that first.

Reply:
Yes, I should resolve the limitation.